Introduce ImGuizmo gem

michalpelka opened the pull request:
It is a gemification of excellent ImGui extension : https://github.com/CedricGuillemet/ImGuizmo

I humbly added o3de API and system component.
**How it look like**

https://github.com/user-attachments/assets/4ebbf064-5c2a-4143-bcf5-b1a78d627ad2

**How to test it?**
- C++ API
- commandline:
```
<14:39:02>     imguizmo_hide (Command)
<14:39:02>     imguizmo_local (Command)
<14:39:02>     imguizmo_show (Command)
<14:39:02>     imguizmo_world (Command)
```
- Script canvas:
![imguizmo](https://github.com/user-attachments/assets/96992ee5-2789-49a9-8514-c871e69ba0a2)

pijaro reviewed (commented):
Consider rearranging the structure and cmake files:

1. Use `FetchContent` to download cpp and h files
2. Build as 3rd party library with proper target
3. Link the 3rd party library with your gem

You can check the [MiniAudio gem](https://github.com/o3de/o3de/tree/development/Gems/MiniAudio) - it follows the same rules and does exactly what want to achieve here.

Nevertheless, fantastic contribution.

michalpelka commented:
@pijaro :
```Use FetchContent to download cpp and h files```
Cannot do that since I've modified a little orginal code (moved some content of headers to API of the Gem) and modified includes to be compatible with o3de's imgui

```Build as 3rd party library with proper target
Link the 3rd party library with your gem
```
It is not clean 3rd-party library like MiniAudio since it is compiled against o3de's ImGui Gem. 
I would leave it as is.

PawelLiberadzki reviewed (changes requested):
There is an issue that we discussed offline - currently it does not work (segfault with attached script, does not work in Game Launcher).

michalpelka commented on Gems/ImGuizmo/Code/Source/Clients/ImGuizmoSystemComponent.h:
```diff
@@ -0,0 +1,125 @@
+            return AZStd::stoi(AZStd::string(arguments[0].data(), arguments[0].size()));
+        };
+
+        static void imguizmo_acquire(const AZ::ConsoleCommandContainer& arguments)
```
I wil change those to follow ImguiGem:
```
<11:51:25> imgui_AutoEnableComponents= [  ]
<11:51:25> imgui_ControllerMouseSensitivity=5 [  ]
<11:51:25> imgui_DiscreteInputMode=1 [  ]
<11:51:25> imgui_EnableAssetExplorer=0 [  ]
<11:51:25> imgui_EnableCameraMonitor=0 [  ]
<11:51:25> imgui_EnableController=0 [  ]
<11:51:25> imgui_EnableControllerMouse=0 [  ]
<11:51:25> imgui_EnableEntityOutliner=0 [  ]
<11:51:25> imgui_EnableImGui=0 [  ]
```

PawelLiberadzki replied:
Great!

michalpelka commented on Gems/ImGuizmo/Code/Source/Clients/ImGuizmoSystemComponent.cpp:
```diff
@@ -0,0 +1,316 @@
+            const auto gizmoTranslation = conversions::Float16ToTranslation3D(gizmoData.m_gizmoMatrix);
+            const AZ::Vector4 localPosition = viewportContext->GetCameraViewMatrix() * AZ::Vector4(gizmoTranslation, 1.0f);
+            // Skip gizmos that are behind the camera
+            if (localPosition.GetZ() > 0.0f)
```
It would be best to do frustum culling here.

michalpelka commented on Gems/ImGuizmo/Code/Source/Clients/ImGuizmoSystemComponent.cpp:
```diff
@@ -0,0 +1,316 @@
+            ImGuizmo::Enable(true);
+            ImGuizmo::SetRect(0, 0, viewportContext->GetViewportSize().m_width, viewportContext->GetViewportSize().m_height);
+
+            const auto gizmoTranslation = conversions::Float16ToTranslation3D(gizmoData.m_gizmoMatrix);
```
Agreed.

PawelLiberadzki reviewed (approved):
Great feature! Works both in game play from Editor and in GameLauncher.
